### PR TITLE
redirect to original location upon successful login

### PR DESCRIPTION
**Pull request by sephiroth-j**
- stores the location before the user is redirected to the login page in the query param "redirect". after successful login, the user gets redirected to this url if it is set. navigation flow will be as follows:
1. http://localhost:8081/projects/285384c9-a18f-465f-921b-d465efd9c43e
2. http://localhost:8081/login?redirect=%2Fprojects%2F285384c9-a18f-465f-921b-d465efd9c43e
3. http://localhost:8081/projects/285384c9-a18f-465f-921b-d465efd9c43e

- uses global navigation guard to ensure the jwt is valid and redirects to login if it is not.
previous implementation checked the jwt only when the app was created. if it was expired afterwards, the app did not recognized that. the guard will take care of this now.

- some other small fixes:
  - "go back" on 404-page always directed the user to "/" --> goes back now as the text implies
  - modal dialog to edit user profile starts fetching the user from the back-end before it already visible. minimizes the effect that the input fields are initially empty and will be filled with content after the dialog is already visible.
  - wrong types for data values in `ProfileEditModal`

references #20

### CODE REVIEW

**Comment by stevespringett** [comment]
@nscuro what impact will this PR have on OIDC flow?

**Comment by nscuro** [comment]
@stevespringett Shouldn't have an impact on OIDC as far as I can tell from just looking at it, other than it simply not working with OIDC in its current state. `oidc-callback.html` would need to be updated to handle the new `redirect` param.

**Comment by nscuro** [comment]
Changing the password (e.g. upon first login with `admin:admin`) doesn't work. The error is: 

```
Uncaught (in promise) Error: Redirected when going from "/login?redirect=%2Fdashboard" to "/change-password" via a navigation guard.
    zt vue-router.esm.js:1958
    Lt vue-router.esm.js:1917
    p vue-router.esm.js:2266
    56d7 index.js:249
    p vue-router.esm.js:2252
    r vue-router.esm.js:1898
    r vue-router.esm.js:1902
    Ft vue-router.esm.js:1906
    confirmTransition vue-router.esm.js:2282
    transitionTo vue-router.esm.js:2152
    replace vue-router.esm.js:2531
    replace vue-router.esm.js:2963
    replace vue-router.esm.js:2962
    login Login.vue:131
    promise callback*login Login.vue:128
    handleSubmit vee-validate.esm.js:1875
    l vee-validate.esm.js:66
    s vee-validate.esm.js:47
    a vee-validate.esm.js:37
    promise callback*l vee-validate.esm.js:39
    o vee-validate.esm.js:40
    o vee-validate.esm.js:36
    handleSubmit vee-validate.esm.js:1865
    submit Login.vue:3
```

**Comment by sephiroth-j** [comment]
@nscuro , thanks for looking at this. I have fixed the error with the "PasswordForceChange" route. The redirection will also work now with this route.

I will have a closer look at the OIDC flow.

**Comment by sephiroth-j** [comment]
> Would it be a viable option to store the redirect target in `localStorage`? As far as I understand it, `redirect` is intendet to be always set by the application and never by the user.

If you open multiple tabs while being logged out and if it stored in `localStorage` or `sessionStorage`, the last opened tab overrides the values for the other tabs. It would also be necessary to remove the value from the storage when it is not needed anymore. Therefore it is better to set the value only then and there, where it is needed.

**Comment by nscuro** [comment]
@sephiroth-j Makes sense, thanks for clarifying.

**Comment by sephiroth-j** [comment]
I have made a few additions to the OIDC flow. @nscuro , would you please test it?

File: src/views/pages/Login.vue [hunk 1]
```diff
@@ -85,6 +85,13 @@ export default {
     ValidationObserver
   },
   data() {
+    let redirectUri = `${ window.location.origin } /static/oidc-callback.html`;
```
Comment (nscuro) [inline src/views/pages/Login.vue:88]: This produces an invalid URL due to the space between `}` and `/static`.

**Comment by nscuro** [comment]
@sephiroth-j Works great, besides the typo in `redirectUri` construction. I'll have to update the OIDC docs though, because IdPs like Keycloak do strict matching against the `redirect_uri` parameter. If `http://dependencytrack.example.com/static/oidc-callback.html` is configured, and `http://dependencytrack.example.com/static/oidc-callback.html?redirect=/vulnerabilities` is provided, the IdP will refuse to authenticate.

File: src/views/pages/Login.vue (hunk 1 above)
Comment (sephiroth-j) [reply src/views/pages/Login.vue]: fixed it

**Comment by sephiroth-j** [comment]
> I'll have to update the OIDC docs though, because IdPs like Keycloak do strict matching against the `redirect_uri` parameter. If `http://dependencytrack.example.com/static/oidc-callback.html` is configured, and `http://dependencytrack.example.com/static/oidc-callback.html?redirect=/vulnerabilities` is provided, the IdP will refuse to authenticate.
Hm, maybe `sessionStorage` is an option here after all.

**Comment by nscuro** [comment]
Shouldn't be an issue. That was supposed to be more like a TODO for me. In Keycloak you can use wildcards for stuff like this, and I'd expect that other IdPs offer similar functionality as well.

**Comment by sephiroth-j** [comment]
I made the checks more restrictive before using the redirect url. Unfortunately, they still don't pass the CodeQL check - I would consider this a false positive.

**Comment by nscuro** [comment]
Looks good to me. Could you include this in 1.3.0 @stevespringett?

**Comment by stevespringett** [comment]
Thank you @sephiroth-j and @nscuro. Really good work here.

**Comment by 37b** [comment]
@nscuro 
> Shouldn't be an issue. That was supposed to be more like a TODO for me. In Keycloak you can use wildcards for stuff like this, and I'd expect that other IdPs offer similar functionality as well.

This is an issue for Azure AD and supposedly the OAuth2.0 spec [rfc6749#section-3.1.2](https://datatracker.ietf.org/doc/html/rfc6749#section-3.1.2) though I think Microsoft's reading of it may be stricter than intended - not sure as this is not my domain. 

Either way, OIDC is only partially working in DTrack with Azure AD as it does not allow wildcards in return URIs. Can the original location be stored elsewhere?

**Pull request merged**